Consistently treat `let` properties as immutable within the type checker

DougGregor opened the pull request:
There are a number of places where references `let` properties are treated as lvalues by the type checker, and we rely on Definite Initialization to detect and diagnose `inout` uses of these properties. This includes:
* `let` instance properties of a type within an initializer.
* Local `let` properties that aren't initialized at point of declaration.

The result of modeling these as lvalues in the type checker is that we could end up selecting a `mutating` function for a `let` base, which would then get diagnosed in Definite Initialization. If there was an alternative non-`mutating` overload that is otherwise not preferred, it will essentially get ignored.

Rework our handling of `let` properties so that they are treated as rvalues uniformly *except* when they are being assigned to. This moves the inout/mutating checking up into the type checker, allowing overload resolution to choose appropriately.

Fixes rdar://127258363, a source compatibility issue with `AsyncIteratorProtocol.next()` that uncovered this issue.

DougGregor committed:
Generalize AbstractStorageDecl::isSettable() to return a three-state value

This operation determines whether a particular storage declaration,
when accessed from a particular location, is mutable or not. It has a
particular semantic that `let` declarations, when accessed from an
initializer, are considered mutable even though they can only be
assigned. There is similar logic for init accessors.

Tease apart "truly mutable" from "initializable because we're in an
initializer", introducing AbstractStorageDecl::mutability() to
represent all three states. isSettable() remains available as a thin
shim over mutability() and all clients are unchanged thus far, making
this a no-op refactoring.

DougGregor committed:
Diagnose inout uses of 'lets' in constructors in the type checker.

Stored `let` properties of a struct, class, or actor permit
'inout' modification within the constructor body after they have been
initialized. Tentatively remove this rule, only allowing such `let`
properties to be initialized (assigned to) and not treated as `inout`.

Fixes rdar://127258363.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test source compatibility

xedin commented on include/swift/Sema/ConstraintSystem.h:
```diff
@@ -2157,6 +2157,11 @@ class ConstraintSystem {
+  /// Expressions that are in the left-hand side of an assignment
+  /// expression. This is used to help distinguish assignments from
+  /// any other potential use that might need lvalues.
+  llvm::SmallPtrSet<Expr *, 2> TargetOfAssignExprs;
```
I think this needs to be controlled by the solver scope because result builder generate assignments as part of the transform which means that if not controlled this can grow pretty large.

xedin replied:
Since the constraint system records parent chains maybe this could be replaced with a walk up the parent chain until it reaches assignment expression?

DougGregor replied:
I missed the fact that we record parent chains; I'd much prefer that to having to keep solver state!

DougGregor replied:
I did this and it's so much better, thank you!

xedin commented on include/swift/Sema/ConstraintSystem.h:
```diff
@@ -4436,7 +4441,10 @@ class ConstraintSystem {
-        });
+        },
+      llvm::function_ref<bool(Expr *)> isAssignTarget = [](Expr *) {
+        return true;
```
Is this the right default?

DougGregor replied:
It's the default that keeps existing code working the same way as it has before. I could remove the default and force all callers to choose.

xedin replied:
Maybe this is going to be a non-issue if you go with walk approach since it could just be a method on a constraint system?

DougGregor replied:
I've cleaned this up and removed the default, thanks for the nudge.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -4242,6 +4242,12 @@ namespace {
+      // Find all of the entries on the left-hand side of an assignment
+      // expression.
+      if (auto assign = dyn_cast<AssignExpr>(expr)) {
+        recordAssignmentTargets(assign->getDest());
```
This looks similar to what happens with argument labels, should probably be moved into `ConstraintGenerator::visitAssignExpr` itself.

DougGregor replied:
This code is gone now, since I'm using parent pointers

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -4307,6 +4313,49 @@ namespace {
     }
 
+    /// Walk the destination of an assignment expression, recording
+    /// each of the
```
Nit: incomplete comment.

DougGregor replied:
(Code is gone)

xedin commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -1434,9 +1434,13 @@ ConstraintSystem::getWrappedPropertyInformation(
+    AbstractStorageDecl *storage, Type baseType,
+    DeclContext *useDC,
+    llvm::function_ref<bool(Expr *)> isAssignTarget = [](Expr *) {
+      return true;
```
Similar to `getUnopenedTypeOfReference` I'm not sure if that's always correct to default to `true`, what if argument is omitted by mistake?

DougGregor replied:
I've removed the defaults, thanks

xedin commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -1448,9 +1452,33 @@ doesStorageProduceLValue(AbstractStorageDecl *storage, Type baseType,
+      if (!memberLocator)
+        break;
+
+      auto *anchor = getAsExpr(memberLocator->getAnchor());
```
Should this me `simplifyToAnchor` instead?

DougGregor replied:
Yes, thanks!

DougGregor commented:
https://github.com/apple/swift-syntax/pull/2653

@swift-ci please test

DougGregor commented:
https://github.com/apple/swift-syntax/pull/2653

@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
[Type checker] Represent "initializable" let instance property references as lvalues

As we do with references to initializable local lets, teach the type
checker to produce ASTs where member references to initializable
instance property lets (e.g., within an initializer) treat the `let`
as an lvalue and then immediately load.

This modeling addresses a source compatibility issue uncovered in
swift-syntax, where code that *technically* has a
use-before-definition on a `let` property in an initializer wasn't
diagnosed as such because the loaded value wasn't actually used for
anything.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please clean smoke test macOS

DougGregor commented:
@swift-ci please clean test source compatibility

DougGregor commented:
@swift-ci please test source compatibility
